### ISSUE TITLE
Why this restriction on Snapshotting?

### ISSUE THREAD

**Issue by otoolep** [hashicorp/raft#617]
From: https://github.com/hashicorp/raft/blob/42d34464b2d203e389e11ed6d43db698792c0604/snapshot.go#L165

"We don't support snapshots while there's a config change outstanding"

Why? I don't follow this. Since the code latest only compacts logs up to the `snapReq.Index` why does it matter? The "state" doesn't have to represent the config change since it will remain in the logs, and is in the Raft DB.

**Comment by banks** [comment]
Hey @otoolep 

We think the logic here is this:

 1. The `SnapshotStore.Create` method currently only supports adding a single (assumed committed) configuration to the snapshot's metadata.
 2. We are using the Raft thesis' simplified configuration change design which assumes that it's safe for a follower to switch to a new configuration as soon as it sees the change in AppendEntries (even before it knows it's committed on a quorum). This is because we only allow single-node config changes which means that the quorums between one configuration and the next always still overlap.
 3. But for that to be correct we have to be careful about 2 things:
     1. We must have at most one uncommitted configuration in flight (enforced by the leader)
     2. We must be able to switch back to the last-known-committed configuration on a follower if we find out the latest configuration we saw actually wasn't committed by a new leader. Otherwise we might be using a configuration that _doesn't_ have fully overlapping quorums with a configuration that a new leader proposed.

So let's suppose we choose to call `SnapshotStore.Create` with only the latest (uncommitted) configuration and persist that. This could violate assumption 2 above because the follower might crash, come back up and _only_ know about the uncommitted configuration but not have any earlier committed configuration. It might then learn from a new leader that the configuration it has was never committed and a new one was chosen in a higher term. Now it's stuck without a valid configuration to continue with. In fact, there's not even any way it would know that configuration in the metadata was uncommitted (with the current design) since we don't have a way to indicate that! So it would have to assume it was committed could potentially split-brain the cluster if it's incompatible with the actual configuration the rest of the cluster agreed.

So the options were:
 1. Change the`Create` interface method to allow passing both the last committed configuration and an option in-flight (uncommitted) one to `Create` and restore both into `configurations.committed` and `configurations.latest`. This would be fine but a bunch of work and a breaking API change.
 2. Just choose not to persist snapshots while in the middle of a config change. Since config changes should be rare and snapshots also infrequent compared with writes in general this seems liked a simple and acceptable strategy. We assume this is why the original authors choose this approach and mostly corresponds with the comment in the code you linked.

> Since the code latest only compacts logs up to the snapReq.Index why does it matter? The "state" doesn't have to represent the config change since it will remain in the logs, and is in the Raft DB.

I think the issue is that the current committed snapshot has probably _already_ been truncated from the logs long ago if it happened 3 months ago on a busy cluster. The only way it's being persisted right now is through the _metadata_ of each snapshot. That metadata currently only supports a single configuration per snapshot so we only store the one committed configuration.

Does that make more sense?

Are you seeing this cause problems?

**Comment by otoolep** [comment]
OK, this is helpful -- thanks. I generally follow what you are saying.

I actually came across this as I am developing a new Snapshot Store for rqlite. When fsm.Snapshot() is called my new Store makes changes to the FSM and also assumes that Persist() will always be called. This code path stood out as one way that Persist() might not be called. 

It's not really a huge deal right now -- this is a fine library as it is, and I'm thinking about how I can deal with this. I had just been making two assumptions that are not actually true:
- If `fsm.Snapshot()` is called then `Persist()` will always be called on the returned snapshot object.
- If `Persist()` is called on the returned Snapshot object, and it returns no error, then the snapshot operation was 100% successful. This is not true either (leaving side a sudden process exit, which can be handled at start-up time). For example, log compaction could fail after `Persist()` was called (unlikely, but still)

Another way of looking at this is this library assumes that a Snapshot generated by the FSM can be thrown away at anytime, without causing the FSM any issues. This is not the case in some recent code I've written. I had to account for this in rqlite just this week -- check out this: https://github.com/rqlite/rqlite/pull/1905

Basically rqlite needs to handle the case where a Snapshot was triggered, but that snapshot state was never persisted to the Snapshot Store. Something like this in the Godocs might be helpful: https://github.com/hashicorp/raft/pull/618
